spec/lib/configuration.spec.js

```diff
@@ -0,0 +1,295 @@
+'use strict';
+
+const Configuration = require('../../lib/configuration').Configuration;
+const CLIOptionsMock = require('../mocks/cli-options');
+
+describe('the Configuration module', () => {
+  let cliOptionsMock;
+
+  beforeEach(() => {
+    cliOptionsMock = new CLIOptionsMock();
+    cliOptionsMock.attach();
+  });
+
+  afterEach(() => {
+    cliOptionsMock.detach();
+  });
+
+  it('overrides default options with customized options', () => {
+    let config = new Configuration({ fromString: true }, { fromString: false });
+    expect(config.getAllOptions().fromString).toBe(true);
+  });
+
+  describe('the getAllOptions() function', () => {
+    it('returns the entire options object', () => {
+      let options = {
+        rev: 'dev & prod',
+        minify: true,
+        inject: { dev: { value: true } }
+      };
+      let config = new Configuration({}, options);
+      expect(config.getAllOptions()).toEqual(options);
+    });
+  });
+
+  describe('the getValue() function', () => {
+    it('supports multi level options', () => {
+      let options = new Configuration({}, {
+        foo: {
+          bar: {
+            dev: {
+              value: 'someValue'
+            },
+            staging: {
+              value: 'someOtherValue'
+            }
+          }
+        }
+      }, 'dev');
+      expect(options.getValue('foo.bar').value).toBe('someValue');
+
+      options = new Configuration({}, {
+        foo: {
+          bar: {
+            'dev & prod': {
+              value: 'someValue'
+            }
+          }
+        }
+      }, 'dev');
+      expect(options.getValue('foo.bar').value).toBe('someValue');
+
+      options = new Configuration({}, {
+        foo: {
+          bar: {
+            dev: {
+              value: {
+                options: true
+              }
+            }
+          }
+        }
+      }, 'dev');
+      expect(options.getValue('foo.bar').value.options).toBe(true);
+
+      options = new Configuration({}, {
+        foo: {
+          bar: 'abcd'
+        }
+      }, 'dev');
+      expect(options.getValue('foo.bar')).toBe('abcd');
+    });
+
+    it('supports one level options', () => {
+      let options = new Configuration({}, { foo: 'someValue' }, 'dev');
+      expect(options.getValue('foo')).toBe('someValue');
```

Comment:
@JeroenVinke the question is for 
```javascript
new Configuration({}, {foo: 'dev'}, 'dev');
new Configuration({}, {foo: 'dev'}, 'stage'); // very tricky
new Configuration({}, {foo: 'dev & stage'}, 'dev'); // this is easier
```
What we want for `getValue('foo')`.

For case#2, it looks quite tricky, we need to known all available env types to determine whether `'dev'` is a condition or plain string.